packit_service_fedmsg/callbacks.py

```diff
@@ -0,0 +1,177 @@
+
+
+def _fedora_dg_push(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    if getenv("PROJECT", "").startswith("packit") and not specfile_changed(
+        event,
+    ):
+        return CallbackResult(
+            msg="[Fedora DG] No specfile change, ignoring the push.",
+            pass_to_service=False,
+        )
+
+    if commit := event.get("commit"):
+        what = f"{commit.get('repo')} {commit.get('rev')}@{commit.get('branch')}"
+    else:
+        what = "Couldn't get commit out of the event."
+
+    return CallbackResult(msg=f"[Fedora DG] Passing push: {what}")
+
+
+def _fedora_dg_pr_flag(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    if nested_get(event, "pullrequest", "user", "name") != packit_user:
+        return CallbackResult(
+            msg=f"[Fedora DG] Flag changed in a PR not created by {packit_user}, ignoring.",
+            pass_to_service=False,
+        )
+
+    return CallbackResult(
+        msg=(
+            "[Fedora DG] Flag on "
+            f"{nested_get(event, 'pullrequest', 'project', 'fullname')}"
+            f" changed to '{nested_get(event, 'flag', 'comment')}'"
+        ),
+    )
+
+
+def _fedora_dg_pr_comment(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    project = nested_get(event, "pullrequest", "project", "fullname")
+    comments = nested_get(event, "pullrequest", "comments")
+    last_comment = comments[-1]
+    return CallbackResult(
+        msg=(
+            f"[Fedora DG] For {project}"
+            f" new comment: '{last_comment['comment']}'"
+            f" from {last_comment['user']['name']}"
+        ),
+    )
+
+
+def _fedora_dg_pr_closed(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    project = nested_get(event, "pullrequest", "project", "fullname")
+
+    if not nested_get(event, "pullrequest", "merged"):
+        return CallbackResult(
+            msg=f"[Fedora DG] Pull request  in {project} was closed, ignoring.",
+            pass_to_service=False,
+        )
+
+    return CallbackResult(
+        msg=f"[Fedora DG] Merged pull request in {project}.",
+    )
+
+
+def _hotness_bugzilla(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    package = event.get("package")
+    version = nested_get(event, "trigger", "msg", "project", "version")
+
+    return CallbackResult(
+        msg=f"[Hotness] New update of package {package} to version {version}.",
+    )
+
+
+def _anitya_version_update(topic: str, event: dict, packit_user: str) -> CallbackResult:
+    package = None
+    for p in nested_get(event, "message", "packages"):
+        if p.get("distro") == "CentOS":
+            package = p.get("package_name")
+            break
+
+    if package is None:
+        return CallbackResult(
+            msg=f"[Anitya] CentOS mapping is not configured for {package}, ignoring.",
+            pass_to_service=False,
+        )
+
+    new_versions = nested_get(event, "message", "project", "upstream_versions")
```

Comment:
how will this work if there are multiple new versions? Could you point me to the docs explaining these, e.g. "upstream_versions" vs "version" (and maybe add a comment with this link please)

Reply:
Very good question, open for discussion IMO :joy: [docs](https://anitya-messages.readthedocs.io/en/latest/topics.html#anitya_schema.project_messages.ProjectVersionUpdatedV2)

I think this applies mainly to the upstreams that keep multiple running branches (like kernel, maybe Java (? even though they have separate RPM packages for that)).

OTOH this is something that I can see used with CentOS specifically, because if there are multiple major/minor releases maintained, user should be able to specify some tag filtering in sense of “I want to follow this branch instead of the latest one”.

Given this [example](https://apps.fedoraproject.org/datagrepper/v2/id?id=290cc3be-6459-4f87-ba3e-d6c879399c68&is_raw=true&size=extra-large), these are the relevant fields:
* `upstream_versions`: which are apparently the **newly** detected versions
* `stable_versions`: duh…
* `versions`: all versions (including stable, newly detected and prereleases; literally all)
* `old_version` (not relevant, but worth mentioning): **latest** from `version`, **CAN** be prerelease

I think we'll need to adjust this in the Service…

Reply:
ok, got it, thanks for the explanation!

> I think we'll need to adjust this in the Service

:+1: